### PR TITLE
Update GHE Interpolation Routine 

### PR DESCRIPTION
**Pull request by mitchute**
Pull request overview
---------------------
 - Updates the GHE interpolation routine

In working on something else, I noticed that the GHE interpolation routine was pretty convoluted and more complicated than it needed to be. This updates it to be more straightforward and to use features from the stdlib. I've also added a few more unit tests. 

I'm seeing small diffs, which I never fully got to the root cause of. When I write the interpolation input/output to a file for the DOAToWaterToAirHPSupply example file and then compare it against the scipy interpolation routine, I get a slightly better RMSE for the updated function. I don't think it's worth digging in any further. Both versions were working, but this one is more modern and is *slightly* more accurate.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 1]
```diff
@@ -2957,99 +2956,40 @@ Real64 GLHESlinky::calcHXResistance(EnergyPlusData &state)
 
 //******************************************************************************
 
-Real64 GLHEBase::interpGFunc(Real64 const LnTTsVal // The value of LN(t/TimeSS) that a g-function
-) const
+Real64 GLHEBase::interpGFunc(Real64 const LnTTsVal) const
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Let's see, this function is only about 100x easier to read and understand now.  This is great.  I might have a few suggested tweaks, but this seems awesome.

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 2]
```diff
@@ -2957,99 +2956,40 @@ Real64 GLHESlinky::calcHXResistance(EnergyPlusData &state)
...
-    //       MODIFIED       August, 2000
-    //       RE-ENGINEERED Dan Fisher
+    // Purpose: interpolate between g-function values, with linear extrapolation above and below range
 
-    // PURPOSE OF THIS SUBROUTINE:
-    //    To interpolate or extrapolate data in GFILE
-    //    to find the correct g-function value for a
-    //    known value of the natural log of (T/Ts)
+    auto &x = this->myRespFactors->LNTTS;
+    auto &y = this->myRespFactors->GFNC;
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Normally, using references to lists is not the right approach, but this is a bit different.  In this case, you don't actually know which index you are going to use yet.  Using `x` and `y` here make this logic **extremely** easy to read and understand.  :+1:

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 3]
```diff
@@ -2957,99 +2956,40 @@ Real64 GLHESlinky::calcHXResistance(EnergyPlusData &state)
...
 
-    //  REFERENCE:          Thermal Analysis of Heat Extraction
-    //                      Boreholes.  Per Eskilson, Dept. of
-    //                      Mathematical Physics, University of
-    //                      Lund, Sweden, June 1987.
+    auto x_begin = x.begin();
+    auto x_end = x.end();
+    double x_val = LnTTsVal;
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc]: Is it worth using references for any of these?  I'm not sure how heavy an iterator is, but usually we use references for anything heavier than POD.  Also the references could be const I think just to make it clear.

File: src/EnergyPlus/GroundHeatExchangers.cc [hunk 4]
```diff
@@ -2957,99 +2956,40 @@ Real64 GLHESlinky::calcHXResistance(EnergyPlusData &state)
...
-                       (this->myRespFactors->GFNC(Mid - 1) - this->myRespFactors->GFNC(Mid)) +
-                   this->myRespFactors->GFNC(Mid);
-
-        return gFuncVal;
-    }
+    double x_low = x[l_idx];
+    double x_high = x[u_idx];
+    double y_low = y[l_idx];
+    double y_high = y[u_idx];
+    return (x_val - x_low) / (x_high - x_low) * (y_high - y_low) + y_low;
```
Comment (Myoldmopar) [inline src/EnergyPlus/GroundHeatExchangers.cc:2992]: These intermediate variables need to be `Real64`.  I was going to say you could also remove them entirely and just use the lookups in the function, but frankly, this makes it just so easy to see the interpolation, that I'd say leave them.

File: tst/EnergyPlus/unit/GroundHeatExchangers.unit.cc [hunk 5]
```diff
@@ -1336,6 +1336,106 @@ TEST_F(EnergyPlusFixture, GroundHeatExchangerTest_System_calcGFunction_UHF)
...
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.491848), -2.133624, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.458741), -2.112028, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.425634), -2.090433, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.392528), -2.068711, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.359421), -2.046238, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.326314), -2.023765, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.293207), -2.001293, tolerance);
+    EXPECT_NEAR(thisGHE.interpGFunc(-14.260100), -1.978820, tolerance);
+}
+
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/GroundHeatExchangers.unit.cc:1438]: Fantastic.

**Review by Myoldmopar — COMMENTED** [review]
Couple incredibly small tweaks, but this is a great improvement.  A 1-D interpolation function shouldn't be all that complex, and this makes that a reality for the GHE.  Could consider generalizing this out later for other parts of E+ that might use this.

**Comment by nealkruis** [comment]
@mitchute, I'm curious if you'd consider using btwxt for the interpolation? It is a generalized n-dimensional interpolation library that is already linked to EnergyPlus and used for tabular interpolation elsewhere in the code. How many dimensions are you interpolating here?

**Comment by mitchute** [comment]
This is just a simple 1D interpolation. We could use btwxt here - I don't have a strong preference either way. I had to roll my own for a separate project, and I had this to hand already so that's why I threw it in.

**Comment by nealkruis** [comment]
That's fair. Btwxt might be overkill if it's always 1-D.

**Comment by Myoldmopar** [comment]
I fully agree that btwxt would be fine here, but also that's for sure overkill here in this always-1-D case.  For now let's forge ahead with this super minimal updated version.

**Comment by mitchute** [comment]
@Myoldmopar comments have been addressed and this is ready for review.

**Review by Myoldmopar — APPROVED** [review]
All happy here now, thanks @mitchute

**Comment by Myoldmopar** [comment]
The tiny diffs are acceptable, but I don't want to muddy other CI results this afternoon, so I'll hold off from merging this until later.  But it's ready.

**Comment by Myoldmopar** [comment]
This one's next, testing locally now.

**Comment by Myoldmopar** [comment]
All happy with latest develop pulled in, merging now, thanks @mitchute

**Pull request merged**